### ISSUE TITLE
Make it possible to compile wasm that can be run without JS e.g. with wasmtime

### ISSUE THREAD

**Issue by liquidaty** [nushell/nushell#14686]
### Related problem

I'd like to build a nu.wasm target that allows me to run e.g.:

`wasmtime nu.wasm script.nu'

But, I have not been able to find any way to build nu.wasm

### Describe the solution you'd like

`cargo build --release --no-default-features --target wasm32-wasi`

or similar

### Describe alternatives you've considered

I've spent quite a bit of time looking at https://github.com/nushell/nushell/pull/14418 and the announcement at https://www.nushell.sh/blog/2024-12-24-nushell_0_101_0.html that says that nushell can be compiled to wasm, but I have not been able to replicate this success, and even if I was, am not sure that it would work with wasmtime.

Tried `cargo build --release --no-default-features --target wasm32-wasi` and `cargo build --release --no-default-features --target wasm32-wasip1` and get numerous errors such as:

```
   ...
   Compiling litemap v0.7.4
error[E0433]: failed to resolve: could not find `unix` in `os`
...
`6 | use std::os::unix::prelude::*;
  |              ^^^^ could not find `unix` in `os`
  ```

Also ran the commands defined in toolkit.nu for "build wasm" and those ran fine, but not sure how that gets me closer to a final nu.wasm binary.

The above was all run on MacBook M3

**Comment by fdncred** [comment]
I think the only currently supported wasm target is wasm32-unknown-unknown and all crates aren't supported. You can see how we do this in the toolkit.nu script. https://github.com/nushell/nushell/blob/5314b31b121615bf851b67dcc2669580d604e220/toolkit.nu#L596-L659

@cptpiepmatz may have more information

**Comment by cptpiepmatz** [comment]
Yeah, you can compile enough crates to have a running nu engine but you don't have the full cli yet and have to build your own tooling (I'm doing something at [cptpiepmatz/nushell-web-demo](https://github.com/cptpiepmatz/nushell-web-demo)).

We use quite a lot of os features with custom handling for Unix and Windows which makes it hard to compile for WASI. Crates like `os-pipe` also don't work for example. I tried to get around it using [WASIX](https://wasix.org) but that feels also very experimental currently.

Then we also have the situation that the target range for WASM is quite complex.
```
nu ❯ rustc --print target-list | lines | where $it =~ "wasm"
╭───┬───────────────────────────╮
│ 0 │ wasm32-unknown-emscripten │
│ 1 │ wasm32-unknown-unknown    │
│ 2 │ wasm32-wasi               │
│ 3 │ wasm32-wasip1             │
│ 4 │ wasm32-wasip1-threads     │
│ 5 │ wasm32-wasip2             │
│ 6 │ wasm64-unknown-unknown    │
╰───┴───────────────────────────╯
```

I don't know if we ever get there but I'm pushing forward as far as I can.

**Comment by liquidaty** [comment]
> I don't know if we ever get there

I can't see any technical reasons preventing it, at least with some limitations that are common to wasm builds.

Even a massively-limited wasm version of nu, that only supported, for example, `let`, `for`, `print`, and basic string/list operations such as `str`, `split`, `each` `join`, would be a helpful start, at least for me. None of those features require any IO, and while pipes would be nice, it isn't necessary so long as basic IO is available (e.g. read a file, write a file using e.g. stock `std::io/std::fs`).

Maybe not enough to be considered having "gotten there", but for my purposes, still extremely useful for real-world application, and a simple start that could then be incrementally expanded.

Has anyone tried anything like the above? If so please share. Or anyone have other suggestions as the shortest path I could take to try to build such a minimal wasm version of nu?

Note: really appreciate the work at [cptpiepmatz/nushell-web-demo](https://github.com/cptpiepmatz/nushell-web-demo); I wasn't able to get it to compile for a fully standalone target that can be run via e.g. `wasmtime` but maybe with just a few adjustments or scope limitations it could?

**Comment by cptpiepmatz** [comment]
> Has anyone tried anything like the above? If so please share. Or anyone have other suggestions as the shortest path I could take to try to build such a minimal wasm version of nu?

If you want, you can try to expand the number of crates that are wasm compatible. https://github.com/nushell/nushell/blob/5314b31b121615bf851b67dcc2669580d604e220/toolkit.nu#L596-L617 In most cases this is solved by feature gating the `os` feature with elements that require os interaction. With enough crates, we should be able to build nushell for wasm.

**Comment by liquidaty** [comment]
> If you want, you can try to expand the number of crates that are wasm compatible.

Does *every* crate need to compile to the target before any runnable binary can be built at all? Would it be feasible to alternatively just shrink the scope of what is included in the build until we have the smallest possible shell that can successfully compile/run? Sorry if this is a dumb question-- I'm a bit new to rust (and nu)

**Comment by fdncred** [comment]
I could guess crates like `nu-cli` wouldn't be needed since it's really mostly setup and repl stuff. Not sure about the others.
